examples/algorithms/include/clusterization/clusterization_algorithm.hpp

```diff
@@ -0,0 +1,82 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "edm/cell.hpp"
+#include "edm/cluster.hpp"
+#include "edm/internal_spacepoint.hpp"
+#include "edm/measurement.hpp"
+#include "edm/spacepoint.hpp"
+#include "geometry/pixel_segmentation.hpp"
+
+// clusterization
+#include "clusterization/component_connection.hpp"
+#include "clusterization/measurement_creation.hpp"
+#include "clusterization/spacepoint_formation.hpp"
+
+namespace traccc {
+
+class clusterization_algorithm
+    : public algorithm<
+          const host_cell_container&,
+          std::pair<host_measurement_container, host_spacepoint_container> > {
+    public:
+    struct config {
+        // currently left empty
+    };
+
+    output_type operator()(const input_type& cells_per_event) const override {
+        output_type o;
+        this->operator()(cells_per_event, o);
+        return o;
+    }
+
+    void operator()(const input_type& cells_per_event,
+                    output_type& o) const override {
+        // output containers
+        auto& measurements_per_event = o.first;
+        auto& spacepoints_per_event = o.second;
+
+        // reserve the vector size
+        measurements_per_event.headers.reserve(cells_per_event.headers.size());
+        measurements_per_event.items.reserve(cells_per_event.headers.size());
+        spacepoints_per_event.headers.reserve(cells_per_event.headers.size());
+        spacepoints_per_event.items.reserve(cells_per_event.headers.size());
+
```

Comment:
To use the OpenMP (if the code was compiled with OpenMP support), the pragmas are needed:
`#if defined(_OPENMP)
#pragma omp parallel for
#endif`
The issue is with the containers that aggregate the results (e.g. `spacepoints_per_event.items.push_back`). They need synchronization mechanisms to ensure the consistency of the result. The open issue is still where would it be better to handle the synchronization? (a) at the user level, i.e. wrap the `container.push_back()` function into a critical region OR (b) at the container level, i.e. the container's implementation of the push_back() function already handles this case. There are pros and cons for each one. Maybe we can discuss more in the next parallelization meeting.

Reply:
Thanks for reply. Maybe you can have a look into `examples/algorithms/cpu/include/clusterization/clusterization_algorithm.hpp` and `examples/algorithms/cpu/include/track_finding/seeding_algorithm` to catch up. This algorithms are applied to `seq_example` but not to `par_example` and `io_dec_par_example` as I didn't decide how to put openmp codes into algorithm files. We can also talk about how to do it in the next meeting

> if we know an approximate number of space points and measurements coming from each cell, we could allocate a fixed region in the container for each cell and in the end, just remove the empty spaces.

I think I've done it in `traccc/examples/algorithms/cpu/include/clusterization/clusterization_algorithm.hpp` by removing `push_back()`.

There are also several things that I want to point out: 
1) In my opinion, we have too many execution files: `seq_example` , `par_example` and `io_dec_par_example`. Once we determine which of either normal container or aggregated one (io_decoupled) is better, we may be able to merge two openmp run files into single one. We also have to decide whether to merge `seq_example` and openmp run file into one or not. It is possible if we put openmp codes into `clusterization_algorithm` or `seeding_algorithm`.
2) When checking the `par_example`, I found that there are two `omp parallel for`: one for events and the other for container. Having this two level parallelizations is better than just parallelizing only for events?

Reply:
The removal of the `push_back()` calls should be fine since `vecmem::jagged_vector` needs to guarantee that the memory addresses of its elements do not overlap, so we should not have race conditions. Maybe @stephenswat can confirm this. 

Regarding the 2 imbricated `parallel` loops: the answer depends on (a) the size of the loops and (b) the amount of computation done in each iteration. For the track fitting part in ACTS, it makes sense to have 2 imbricated loops (it reduced the memory consumption by half without impacting the wall clock execution time). If we have a smaller use case here, we can remove the container level loop. Do you have some estimates on the container sizes and maybe a rough estimate of the FLOPS done inside the loop?

I am happy to merge several examples if needed. Maybe we can discuss more in the next meeting.

Reply:
container size is the number of modules so I think it is O(10^3) for tml detector and O(10^4) for atlas detector, respectively. And I have no idea about FLOPS.

> I am happy to merge several examples if needed

Yup. But I think it is okay to leave them until we have a complete tracking chain

Reply:
Instead, can you consider updating `par_example` and `io_dec_par_example` as a follow-up PR?